### DISCUSSION TITLE
运行例程时报 parse 错误

### DISCUSSION THREAD

**Discussion by nobodxbodon** [yuyan-lang/yuyan#72] [Q&A] [answered]
```
% ./yy examples/example.yuyan
[2024-01-17 14:01:02] cfp : [examples/example.yuyan] Updated Content success: true updated: true
[2024-01-17 14:01:02] cfp : [examples/example.yuyan] Updated Preprocessing success: false updated: true
examples/example.yuyan:3:1 - 3:5:错误(error)：在理解(parse)`阳为爻也`时出现问题：不能够理解(parse)输入
阳为爻也。
~~~~~~~~
调试信息：
所有可能的名称(all unknown ids)：阳为爻
所有相关的操作符(all relevant ops)：〇也〇r250
所有（包括不相关的）操作符(all ops)：〇之〇r710，〇合〇r385，化〇而〇r350，化〇者囗而〇r350，承〇者囗而〇r350，承囗而〇r350，〇其实〇n480，〇中的〇r710，〇结合〇r385，从〇到〇r350，从〇的囗到〇r350，自〇的囗到〇r350，自囗到〇r350，〇实际上是〇n480，〇中〇l700，〇于〇l690，〇与〇r680，有〇则〇n640，〇或〇r630，鉴〇而〇r620，若〇则〇否则〇r645，〇授以〇l690，会囗而〇r520，受囗而〇r520，《《C调用》》名〇传〇r530，遇〇者囗而〇r520，循囗以〇r510，虑〇n525，虑囗者〇而〇r526，〇；〇r535，〇也〇r250，〇中的第〇个l700，〇使用于〇l690，〇连结〇r680，如果是〇那么〇n640，〇或者〇r630，分析〇随后〇r620，如果〇那么〇否则〇r645，〇给予〇l690，遇到了囗随后〇r520，得到了囗随后〇r520，《《C调用》》函数名是〇参数是〇r530，遇到了〇的囗随后〇r520，递归囗随后〇r510，执行如下计算〇n525，让囗为〇随后〇r526，类型为〇的〇r250
```

**Comment by UltimatePea** [comment, marked as answer]
examples文件夹的内容已经过时了，我预计等beta版本做完之后才有时间更新，目前例子的话得从 ./yylib ./test ./豫言编译器 这三个文件夹里找。

**Reply by UltimatePea** [reply]
example/example.yuyan 是一个快排程序，最新版在 https://github.com/yuyan-lang/yuyan/blob/master/performance-investigation/quick-sort/quick-sort.yuyan

**Reply by nobodxbodon** [reply]
用yy的话：
```
../yuyan/performance-investigation/quick-sort/quick-sort.yuyan:30:8 - 30:12:错误(error)：`结束时间`类型不匹配(type mismatch) 
 推断的类型(synthesized type) : (real) 
 检查的类型(checked type) : (int) 
 原因：(real) 不等于 (int)
总用时纳秒数者结束时间减开始时间也。
              ~~~~~~~~
在推断`ap(ap(〇减〇, 结束时间), 开始时间)`的类型时
在推断`ap(〇减〇, 结束时间)`的类型时
在检查`结束时间`拥有类型`(int)`时
当前结构名：quick-sort
当前已定义的值及其类型：
结束时间：小数( >>>> ) (定义) >>> (...)；
快排：(ap(〇列( >>>> )⟦...⟧, 整数( >>>> )) -> ap(〇列( >>>> )⟦...⟧, 整数( >>>> ))) (定义) >>> (...)；
《《元变量1494》》：(Set)(resolved metavar >>> (int))；
《《元变量1493》》：(Set)(resolved metavar >>> (int))；
《《元变量1492》》：(Set)(resolved metavar >>> (int))；
《《元变量1491》》：(Set)(resolved metavar >>> (int))；
《《元变量1490》》：(Set)(resolved metavar >>> (int))；
《《元变量1488》》：元类型( >>>> )(resolved metavar >>> (int))；
《《元变量1487》》：元类型( >>>> )(resolved metavar >>> (int))；
《《元变量1485》》：元类型( >>>> )(resolved metavar >>> (int))；
以及之后的240个值或类型
```

用 yy_bs：
```
...
[2024-01-17 15:07:23] | 代码生成准备完成： .../yuyan-0.1.0alpha2.0.2/yylib/标准库/语言核心/内建类型。豫
uv_spawn error: (custom) llvm-as:  no such file or directory
spawn error
```

**Reply by UltimatePea** [reply]
这个文件也过时了， 我更新了一下：

https://github.com/yuyan-lang/yuyan/blob/continue_to_compilation/performance-investigation/quick-sort/quick-sort.yuyan

**Reply by nobodxbodon** [reply]
yy:
```
！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！！
豫言运行环境(yy_runtime)：未捕捉的异常(Uncaught Exception)：
尝试打印值：（可能会出现segfault）：
期待一个参数
```

**Reply by UltimatePea** [reply]
这个就是编译好了，可以运行了。这个程序需要一个命令行参数。

我记得是用 -o 可以指定可执行文件的名称。

**Reply by UltimatePea** [reply]
uv_spawn error: (custom) llvm-as:  no such file or directory
spawn error

这个问题的话需要 brew link llvm

**Reply by nobodxbodon** [reply]
`% brew link llvm --force`之后，yy报错：

```
[2024-01-17 16:23:19] cfp : [../yuyan/performance-investigation/quick-sort/quick-sort.yuyan] Computed LLVMInfo succes : true updated: true
(make -C runtime/ debug_old;clang /Users/xuanwu/work/各种语言/yuyan-0.1.0alpha2.0.2/.yybuild.nosync/yy30583.ll /Users/xuanwu/work/各种语言/yuyan-0.1.0alpha2.0.2/runtime/libyyrtdebug_old.a -g  -o /Users/xuanwu/work/各种语言/yuyan-0.1.0alpha2.0.2/.yybuild.nosync/yyexe  -L /opt/homebrew/lib -l gc -l uv -Wno-override-module)
make: Nothing to be done for `debug_old'.
ld: library not found for -lSystem
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ERROR in Making Executable
sh: /Users/xuanwu/work/各种语言/yuyan-0.1.0alpha2.0.2/.yybuild.nosync/yyexe: No such file or directory
```

yy_bs：
```
尝试打印值：（可能会出现segfault）：
命令执行失败：
clang -o /Users/xuanwu/work/各种语言/yuyan-0.1.0alpha2.0.2/.yybuild.nosync/yy_quick-sort_豫言编译器默认执行包.exe -O3 /Users/xuanwu/work/各种语言/yuyan-0.1.0alpha2.0.2/.yybuild.nosync/yy_quick-sort_豫言编译器默认执行包.opt.bc ./runtime/libyyrtdebug.a -L /usr/local/lib -l gc -l uv -Wno-override-module -g -Werror
标准输出：
错误输出：ld: library not found for -lgc
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

**Reply by UltimatePea** [reply]
哦哦，是这样，Mac 本身自带了一套LLVM 工具链，但只包括clang 和 ld, 不包括llvm-as，我也不理解为何clang会报错


解决方案是这样的：
yy 之前已经编译成功了，但force link之后会出问题，要unlink，然后用系统提供的clang就应该可以成功

yy_bs 之前失败是因为缺少llvm-as 和 opt两个软件，但force link会错误地链接clang，需要brew info llvm

==> llvm: stable 17.0.6 (bottled), HEAD [keg-only]
Next-gen compiler infrastructure
https://llvm.org/
/usr/local/Cellar/llvm/16.0.6 (6,764 files, 1.6GB)
  Poured from bottle using the formulae.brew.sh API on 2023-08-08 at 22:33:02
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/l/llvm.rb
License: Apache-2.0 with LLVM-exception
==> Dependencies
Build: cmake ✘, ninja ✘, swig ✘
Required: python@3.12 ✘, z3 ✔, zstd ✔
==> Options
--HEAD
	Install HEAD version
==> Caveats
To use the bundled libc++ please add the following LDFLAGS:
  LDFLAGS="-L/usr/local/opt/llvm/lib/c++ -Wl,-rpath,/usr/local/opt/llvm/lib/c++"

llvm is keg-only, which means it was not symlinked into /usr/local,
because macOS already provides this software and installing another version in
parallel can cause all kinds of trouble.
==> Analytics
install: 41,145 (30 days), 147,223 (90 days), 388,720 (365 days)
install-on-request: 16,864 (30 days), 75,147 (90 days), 274,066 (365 days)
build-error: 447 (30 days)

找到它的install path，我这里是 /usr/local/Cellar/llvm/16.0.6 ， 这个path底下应该会有一个叫bin的文件夹，要把那个文件夹路径添加到PATH的环境变量里，这样yy_bs就可以正确地调用clang和 llvm-as了。

**Reply by UltimatePea** [reply]
ld: library not found for -lgc 这个错误的话就要把

clang -o /Users/xuanwu/work/各种语言/yuyan-0.1.0alpha2.0.2/.yybuild.nosync/yy_quick-sort_豫言编译器默认执行包.exe -O3 /Users/xuanwu/work/各种语言/yuyan-0.1.0alpha2.0.2/.yybuild.nosync/yy_quick-sort_豫言编译器默认执行包.opt.bc ./runtime/libyyrtdebug.a -L /usr/local/lib -l gc -l uv -Wno-override-module -g -Werror


里面的 -L /usr/local/lib 换成libgc和libuv的安装路径了，然后重新执行clang命令就可以得到可运行程序

**Reply by nobodxbodon** [reply]
unlink了，我先跑yy吧。
该用什么命令行参数呢？随便用了个14好像报错：
```
% ./yy ../yuyan/performance-investigation/quick-sort/quick-sort.yuyan 14
...
[2024-01-17 19:50:50] cfp : [../yuyan/performance-investigation/quick-sort/quick-sort.yuyan] Computed LLVMInfo succes : true updated: true
unhandled exception: Io: openIn "/Users/xuanwu/work/各种语言/yuyan-0.1.0alpha2.0.2/14" failed with SysErr: No such file or directory [noent]
```

**Reply by nobodxbodon** [reply]
-c -o 生成 .out 之后，运行成功：
```
% ./quick-sort.out 14
0.000037
```

**Reply by nobodxbodon** [reply]
装llvm最后提示了：
```
If you need to have llvm first in your PATH, run:
  echo 'export PATH="/opt/homebrew/opt/llvm/bin:$PATH"' >> ~/.zshrc
```
刚加上 PATH，正找哪里改 yy_bs 的 -L

**Reply by UltimatePea** [reply]
```
 $ grep -rIn "/usr/local/lib" ./豫言编译器/ ./src
./豫言编译器/编译步骤/代码生成/低级虚拟机/主程序生成。豫:161:                ，『/usr/local/lib』
./src/build-system/compilation-manager.sml:162:            ^ " -L /usr/local/lib -l gc -l uv "

```

**Reply by nobodxbodon** [reply]
将 `主程序生成。豫`中的路径改为 `/opt/homebrew/lib` 后，再运行 make yy_bs：

```
...
DEBUG : 编译步骤之总体过程之单一编译工作 mapped to 2820550
[2024-01-18 08:51:09] cfp : [豫言编译器/入口。豫] Computed LLVMInfo succes : true updated: true
(make -C runtime/ debug_old;clang /Users/xuanwu/work/各种语言/yuyan-0.1.0alpha2.0.2/.yybuild.nosync/yy2828501.ll /Users/xuanwu/work/各种语言/yuyan-0.1.0alpha2.0.2/runtime/libyyrtdebug_old.a -g  -o /Users/xuanwu/work/各种语言/yuyan-0.1.0alpha2.0.2/yy_bs  -L /opt/homebrew/lib -l gc -l uv -Wno-override-module)
make[1]: Nothing to be done for `debug_old'.
ld: library not found for -lSystem
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ERROR in Making Executable
```

**Reply by UltimatePea** [reply]
是这样，似乎Homebrewd的clang有点问题，这个clang要用系统的clang，不能用brew的llvm clang

(make -C runtime/ debug_old;clang /Users/xuanwu/work/各种语言/yuyan-0.1.0alpha2.0.2/.yybuild.nosync/yy2828501.ll /Users/xuanwu/work/各种语言/yuyan-0.1.0alpha2.0.2/runtime/libyyrtdebug_old.a -g  -o /Users/xuanwu/work/各种语言/yuyan-0.1.0alpha2.0.2/yy_bs  -L /opt/homebrew/lib -l gc -l uv -Wno-override-module)

网上应该有很多解决方案
https://www.google.com/search?q=ld%3A+library+not+found+for+%E2%80%9ClSystem%E2%80%9D

**Reply by nobodxbodon** [reply]
[如此](https://stackoverflow.com/a/70676858) 加了lib路径到 ~/.zshrc，再make yy_bs：
```
[2024-01-18 12:16:25] cfp : [豫言编译器/入口。豫] Computed LLVMInfo succes : true updated: true
(make -C runtime/ debug_old;clang /Users/xuanwu/work/各种语言/yuyan-0.1.0alpha2.0.2/.yybuild.nosync/yy2828501.ll /Users/xuanwu/work/各种语言/yuyan-0.1.0alpha2.0.2/runtime/libyyrtdebug_old.a -g  -o /Users/xuanwu/work/各种语言/yuyan-0.1.0alpha2.0.2/yy_bs  -L /opt/homebrew/lib -l gc -l uv -Wno-override-module)
make[1]: Nothing to be done for `debug_old'.
ld: warning: dylib (/opt/homebrew/lib/libgc.dylib) was built for newer macOS version (14.0) than being linked (11.3)
ld: warning: dylib (/opt/homebrew/lib/libuv.dylib) was built for newer macOS version (14.0) than being linked (11.3)
ld: can't write output file: /Users/xuanwu/work/各种语言/yuyan-0.1.0alpha2.0.2/yy_bs for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
ERROR in Making Executable
```

**Reply by UltimatePea** [reply]
可能要先删除yy_bs

类似的问题：

https://stackoverflow.com/questions/73048687/macos-flutter-ld-cant-write-output-file-for-architecture-x86-64

**Reply by nobodxbodon** [reply]
删了原本的 yy_bs 后再make：
```
error: cannot create Plist: Permission deniedclang: error: dsymutil command failed with exit code 1 (use -v to see invocation)
ERROR in Making Executable
```
大概哪儿还要先清理？

**Reply by UltimatePea** [reply]
这个问题我也没遇到过，可能是权限问题，可以尝试`sudo`来执行clang代码，不过这个操作比较危险，这个是ChatGPT给出的答案。

<img width="695" alt="image" src="https://github.com/yuyan-lang/yuyan/assets/9295852/0160b9fd-ba40-47a2-824e-e24e1713c1f3">

**Reply by UltimatePea** [reply]
还有一个可能是可以检查一下当前文件夹的owner是不是current user

**Reply by UltimatePea** [reply]
总的来说是要找出为什么dsymutil没有权限在想要的路径创建Plist文件

**Reply by nobodxbodon** [reply]
hmm 发现虽然报错但那时还是生成了 yy_bs，但 .dSYM 是老的：
<img width="423" alt="Screenshot 2024-01-18 at 2 06 45 PM" src="https://github.com/yuyan-lang/yuyan/assets/392497/9e8c8920-3419-45d5-abc8-4e2767b300b0">
于是 `% sudo rm -rf yy_bs.dSYM` 之后再 make yy_bs，就无错编过了。
